`PoolConfig` no longer keeps a reference to the connection class (attempt 3)

flavorjones opened the pull request:
### Motivation / Background

Alt to https://github.com/rails/rails/pull/54350 and https://github.com/rails/rails/pull/54349, fixes https://github.com/rails/rails/issues/54343


### Detail

This is variation on #54349 that extends `ConnectionHandler::StringConnectionName` to track `primary_class?`, and introduces `ActiveRecord::Base.preventing_writes?` that will accept a connection name and do the same thing that `current_preventing_writes` does on a record class.


### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [-] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

djmb commented on activerecord/lib/active_record/core.rb:
```diff
@@ -201,6 +201,15 @@ def self.current_preventing_writes
+      def self.preventing_writes?(class_name)
+        connected_to_stack.reverse_each do |hash|
+          return hash[:prevent_writes] if !hash[:prevent_writes].nil? && hash[:klasses].include?(Base)
+          return hash[:prevent_writes] if !hash[:prevent_writes].nil? && hash[:klasses].any? { |klass| klass.name == class_name }
```
We'd need to do the same check in `current_preventing_writes`, `current_role` and `current_shard`

flavorjones replied:
Are you sure about that? The essential part of the bug described in #54343 is that `current_preventing_writes` was being called on the class cached in `PoolConfig`. 

There were no other methods being called on that class, which made it possible to replace it with a modified `StringConnectionName` instance. I believe the existing `current_*` methods in `core.rb` are fine because the classes being stored in the `connected_to_stack` are ephemeral (per-request) and not cached.

byroot commented on activerecord/lib/active_record/core.rb:
```diff
@@ -201,6 +201,15 @@ def self.current_preventing_writes
         false
       end
 
+      def self.preventing_writes?(class_name)
```
```suggestion
      def self.preventing_writes?(class_name) # :nodoc: 
```

This is purely internal API right?

byroot commented on activerecord/lib/active_record/connection_adapters/pool_config.rb:
```diff
@@ -41,14 +40,19 @@ def server_version(connection)
+          @connection_class = connection_class
         else
-          connection_class.name
+          @connection_class = ConnectionHandler::StringConnectionName.new(connection_class.name, connection_class.primary_class?)
```
There's just one minor thing that bother me a bit, it is that now `@connection_class` is always a `StringConnectionName` instance, hence the name makes little sense.


Feels like we should refactor further. AFAIK this is all private API.

flavorjones replied:
I agree. I'm trying to craft a test that reproduces the issue, and then I will refactor and clean up. I appreciate the feedback!

flavorjones committed:
PoolConfig no longer keeps a reference to the connection class.

Keeping a reference to the class caused subtle issues when combined with reloading in
development. Instead, extend StringConnectionName to properly track
`primary_class?` and cache that.

Refactoring includes renaming `connection_class` to
`connection_descriptor`, and renaming the StringConnectionName class
to ConnectionDescriptor.

Fixes #54343.

rafaelfranca commented:
Backported to 8-0-stable since this is mostly an internal restructure to fix a bug
